clang/lib/Sema/SemaARM.cpp

```diff
@@ -569,34 +569,39 @@ static bool checkArmStreamingBuiltin(Sema &S, CallExpr *TheCall,
   // * When compiling for SVE only, the caller must be in non-streaming mode.
   // * When compiling for both SVE and SME, the caller can be in either mode.
   if (BuiltinType == SemaARM::VerifyRuntimeMode) {
-    llvm::StringMap<bool> CallerFeatureMapWithoutSVE;
-    S.Context.getFunctionFeatureMap(CallerFeatureMapWithoutSVE, FD);
-    CallerFeatureMapWithoutSVE["sve"] = false;
+    llvm::StringMap<bool> CallerFeatures;
+    S.Context.getFunctionFeatureMap(CallerFeatures, FD);
 
     // Avoid emitting diagnostics for a function that can never compile.
-    if (FnType == SemaARM::ArmStreaming && !CallerFeatureMapWithoutSVE["sme"])
+    if (FnType == SemaARM::ArmStreaming && !CallerFeatures["sme"])
       return false;
 
-    llvm::StringMap<bool> CallerFeatureMapWithoutSME;
-    S.Context.getFunctionFeatureMap(CallerFeatureMapWithoutSME, FD);
-    CallerFeatureMapWithoutSME["sme"] = false;
+    const auto FindTopLevelPipe = [](const char *S) {
+      unsigned Depth = 0;
+      unsigned I = 0, E = strlen(S);
+      for (; I < E; ++I) {
+        if (S[I] == '|' && Depth == 0)
+          break;
+        if (S[I] == '(')
+          ++Depth;
+        else if (S[I] == ')')
+          --Depth;
+      }
+      return I;
+    };
+
+    const char *RequiredFeatures =
+        S.Context.BuiltinInfo.getRequiredFeatures(BuiltinID);
+    unsigned PipeIdx = FindTopLevelPipe(RequiredFeatures);
+    assert(PipeIdx != 0 && PipeIdx != strlen(RequiredFeatures) &&
+           "Expected feature string of the form 'SVE-EXPR|SME-EXPR'");
+    StringRef NonStreamingBuiltinGuard = StringRef(RequiredFeatures, PipeIdx);
+    StringRef StreamingBuiltinGuard = StringRef(RequiredFeatures + PipeIdx + 1);
```

Comment:
I'm not sure if this is going to change in one of the future patches, but could something like this work here instead?
```
std::string RequiredFeatures =
    str(S.Context.BuiltinInfo.getRequiredFeatures(BuiltinID));
std::smatch match;
if (std::regex_match(str, match, std::regex("(s[^|]+)\\|(s.+)"))) {
  NonStreamingBuiltinGuard = StringRef(match[1].str());
  StreamingBuiltinGuard = StringRef(match[2].str());
}
```

Reply:
This code comes from https://github.com/llvm/llvm-project/pull/145941, which I've now added you to.

I'm not too familiar with `std::regex` so please forgive my ignorance but it seems to only support simple cases and specifically cannot handle nested brackets?  The example given would fail when supplied with `sve(sve2|sve-hh)|sme(sve2|sme2)`?

Reply:
My mistake, I was only considering the simpler cases such as `sve2p1|sme2` so I don't think regex is going to help here. I've approved the other patch which added these changes :)